Update to catbox 10

marcuspoehls opened the pull request:
This PR includes the following updates

- support for `catbox` 10
- refactor library to use an ES2015 `class`
- use `async/await` with promises from MongoDB's lib
- update tests to support changed library code
- bump dependencies to their latest versions
- requires Node.js `8.9.0` or later, defined in `package.json`
- update `.travis.yml` to test Node.js versions 8 and 9

Solves https://github.com/hapijs/catbox-mongodb/issues/27 (catbox 10 support)

marcuspoehls committed:
update to catbox10

jardakotesovec commented:
Great, thanks! Will look into it this weekend.

marcuspoehls commented:
Cool, thank you

jardakotesovec commented:
@marcuspoehls Looks good. Eventually I think it would make sense instead of using new Promise(), to use mongodb promised based api. Basically if you don't pass callbacks mongodb functions will return promises. Would you consider updating that?

marcuspoehls commented:
@jardakotesovec Thank you for the review. Yeah, I’d like to change the MongoDB promise wrappers to native promises.

I checked the MongoDB lib’s readme for promise support, didn’t find anything and then went away. Should have checked the code 😃 Thanks for the hint!

Will update the PR, latest on Monday.

marcuspoehls commented:
@jardakotesovec Updated the lib's code to Use async/await with promises from the MongoDB library. Also: the lib is an ES2015 class now.

Of course the tests are up to date with the lib's changes 😃

The pending Travis check seems to be a display issue with GitHub. Travis already finished the test run successfully.

What do think about the changes?

jardakotesovec commented on lib/index.js:
```diff
@@ -16,204 +16,175 @@ internals.defaults = {
-        if (err) {
-            return connected(new Error('Failed opening connection:' + JSON.stringify(err)));
+        // Check if start already pending
+        if (this.startPending) {
```
In original implementation it was possible to call `start` multiple times and callback was called for all of them once connected. But this looks like it will just return promise immediately when start is called second time instead of wait for actual connection.

@hueniverse Do I assume correctly that the catbox clients should still handle calling `start` multiple times, which all should resolve in same promise once actual connection promise get resolved?

hueniverse replied:
Yes. The cache must be full connected when `start()` resolves no matter how many times it is called. One easy way to do this is to hold on to the promise and return it directly if one has been saved. When you stop the client, you clear out this promise.

jardakotesovec commented on lib/index.js:
```diff
@@ -16,204 +16,175 @@ internals.defaults = {
+            return collection;
+        }
+        catch (err) {
+            throw err;
```
I have not used the await yet as much. But does make sense to actually rethrow here like that? If you would remove the try catch here - would it make any difference?

hueniverse replied:
Just let it go without catching. No need to catch just to throw the same thing again.

jardakotesovec commented on test/index.js:
```diff
@@ -21,334 +21,275 @@ const expect = Code.expect;
 
-        start();
-        start();
+        await start();
```
If you use await here - it won't call start twice at the same event as this test intend to.

jardakotesovec commented on test/index.js:
```diff
@@ -21,334 +21,275 @@ const expect = Code.expect;
-            });
+            await client.start();
+            expect(client.isReady()).to.equal(true);
+            --x;
```
This does need to be here anymore, does it?

jardakotesovec commented on test/index.js:
```diff
@@ -21,334 +21,275 @@ const expect = Code.expect;
-                setTimeout(() => {
-
-                    client.get(key, (err, result) => {
+            setTimeout(async () => {
```
I suspect that the lab won't wait until this setTimeout is evaluated. Since test function basically ends when returned promise gets resolved.

jardakotesovec replied:
To make these patterns nicely consistent within the hapi ecosystem - its often useful to look at how its approached in hapi core module. For example in this case - you can see one how the setTimeout can be used in async function - https://github.com/hapijs/hapi/blob/286324dcd61733148e3759f325659eeffe124184/test/request.js#L265

Marsup replied:
That, or `hoek.wait`.

jardakotesovec commented on test/index.js:
```diff
@@ -563,69 +486,69 @@ describe('Mongo', () => {
         });
 
-        it('sets isReady to true when the connection succeeds', (done) => {
+        it('connects with pending start', async () => {
```
Out of the curiosity, what made you to create this test?

jardakotesovec commented on test/index.js:
```diff
@@ -563,69 +486,69 @@ describe('Mongo', () => {
+            }
         });
 
-        it('calls any pending callbacks waiting for a start', (done) => {
```
I think this should stay - as commented in start function its still important that start function resolves when actual connection starts even if called multiple times before connection is created.

marcuspoehls replied:
Yep, agree. With your explanation I see the need for it

jardakotesovec commented on test/index.js:
```diff
@@ -710,136 +628,124 @@ describe('Mongo', () => {
 
-            mongo.get('test', (err) => {
-
+            try {
```
I suspect that if mongo.get does not throw, it still would pass, because expects in catch would not get called. There is multiple tests like this. Not sure whats the best practice for these tests.

@Marsup Any hint? :-)

Marsup replied:
Throw something different on the new line, or return in the catch and throw after.

marcuspoehls replied:
Alright, thank you for pointing out a solution 👍

jardakotesovec commented:
@marcuspoehls Thanks for updates, looks good. I also learned something about await and async :-).

I want to do one more detailed review to make sure I am leaving this module at least at the same shape as I got it :-). I will get on it likely on Sunday.

I suspect that whats not handled in compared to original implementation is that if the connection fails - you can't really call start again - you will keep returning rejected promise instead of start connecting again. In original implementation it would start connection again I believe.  Have not checked that in detail..  Would be cool to cover that with test.

marcuspoehls commented:
@jardakotesovec Hey Jarda, thank you for letting me know about your review plans.

The current solution with the connection promise isn't ideal. I was thinking about it, but can't find a better solution. Your idea makes sense, I guess the rejected promise can only be delted when closing the connection and opening it again. 

A test is a good idea! Will try to come up with one in the next days.

jardakotesovec commented:
@marcuspoehls It could also be cleaned up in `catch` when the connection fails.

marcuspoehls commented:
@jardakotesovec Do you have an idea how to let the MongoDB connection fail in the first run and succeed in the second try?

jardakotesovec commented:
@marcuspoehls if we don't want to introduce some more complicated mocking [solution](https://github.com/thlorenz/proxyquire). You could alter mongo.settings.uri to correct one before calling start second time.

marcuspoehls commented:
@jardakotesovec Thank you! With every iteration, I learn new aspects of testing and thinking out of the box to cover a lot of situations.

Added the test to check whether it’s possible to connect successfully after a failed connection. Used your approach and override the MongoDB uri.

jardakotesovec commented on test/index.js:
```diff
@@ -21,334 +21,297 @@ const expect = Code.expect;
             expect(err).to.exist();
-            done();
-        });
+        }
```
If client.set is successful the test will pass. Maybe better pattern is to use expects().to.reject() for these scenarios. Like this:
https://github.com/hapijs/catbox-memory/commit/b89f356972116cd18f149e90e0a6cc08e036da22#diff-910eb6f57886ca16c136101fb1699231R165

jardakotesovec replied:
You have this pattern on several places, please review also these other cases.

marcuspoehls replied:
Thank you for pointing me to `expect().to.reject()`. Didn’t know it before 👍

jardakotesovec commented on test/index.js:
```diff
@@ -563,69 +508,64 @@ describe('Mongo', () => {
+            }
+            catch (err) {
                 expect(err).to.not.exist();
                 expect(mongo.isReady()).to.be.true();
```
You are testing isReady on place where it never going to be called. This test could look something like this:
```
await mongo.start()
expect(mongo.isReady()).to.be.true();
```
If the mongo.start would reject instead of resolve, this test would end up with error, so no need always for explicit catch.

marcuspoehls replied:
Good catch! 

You’re right, the test fails `mongo.start()` rejects and therefore the test fails.

jardakotesovec commented on test/index.js:
```diff
@@ -935,62 +832,94 @@ describe('Mongo', () => {
         });
 
-        it('passes an error to the callback when there is an error returned from setting an item', (done) => {
+        it('passes an error to the callback when there is an error returned from setting an item', async () => {
```
I noticed that you renamed many of the tests so it correctly describes that it does - but there is still several old ones talking in callback terms.

marcuspoehls commented:
I updated the tests and removed unnecessary try-catch blocks. Also, no need to catch an error and expect it to exist when there’s `expect().to.reject()`. Removed them as well.

Update for the `lib` that now follows `catbox-memory` with `throw new Boom('message')`.

jardakotesovec commented:
@marcuspoehls Great, I think its time to merge it. Could you squash it to one commit with rebase?

Marsup commented on .travis.yml:
```diff
@@ -1,8 +1,8 @@
-  - "4"
-  - "6"
+  - "8"
+  - "9"
```
Missing "node".

marcuspoehls commented:
@jardakotesovec @Marsup rebased and squashed all commits to a single one. Also added `"node"` to the Node.js version list in `.travis.yml`

Marsup commented:
@marcuspoehls to fail early when a new node release is out.

marcuspoehls commented:
@Marsup ah, I see. Thanks
